Enable image by owner and name

gustavomcarmo opened the pull request:
If the ami code is not known upfront, this feature enables getting the ami code by image's owner and name.

#### PR Type

- Feature Pull Request

themr0c commented:
This is a nice improvement. Would you like to add some example in the documentation? (`molecule/driver/ec2.py`)

decentral1se commented on molecule/cookiecutter/scenario/driver/ec2/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml:
```diff
@@ -59,10 +59,19 @@
       ec2:
         key_name: "{{ keypair_name }}"
-        image: "{{ item.image }}"
+        image: "{{ item.image if item.image is defined else ami_facts.results[index].images[0].image_id }}"
```
Could we drop this onto multiple lines?

gustavomcarmo replied:
For sure!

decentral1se replied:
:+1:

decentral1se commented on molecule/cookiecutter/scenario/driver/ec2/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml:
```diff
@@ -59,10 +59,19 @@
         mode: 0600
       when: keypair.changed
 
+    - name: Get the ec2 ami(s) by owner and name, if image not set
```
Can you also please add this to `test/resources/playbooks/ec2/create.yml` for our functional testing?

gustavomcarmo replied:
In which AWS region do the tests run? The AWS_REGION environment variable is set in Travis CI, right?

decentral1se replied:
They don't run at all :sweat_smile: 

We'll figure it out when we get there, just set whatever for now :)

gustavomcarmo replied:
I've just pushed the required change for functional testing.

Wouldn't be better defining the region directly in the create playbook?

gustavomcarmo commented:
> This is a nice improvement. Would you like to add some example in the documentation? (`molecule/driver/ec2.py`)

I'll do it!

gustavomcarmo commented:
@decentral1se [Here](https://github.com/codeyourinfra/codeyourinfra/blob/98a254e90788aa4d69f2420c4e85c2eefb37fd7e/check_server_log_files/molecule/aws/create.yml#L61) is my own usage of the feature. Clone the repo and follow the steps explained in the README.

gustavomcarmo commented:
@decentral1se Some logs from my project:

```
    TASK [Get the image ami] *******************************************************
    ok: [localhost] => (item={'name': 'jenkins', 'region': 'eu-west-2', 'image_owner': 334305766942, 'image_name': 'codeyourinfra/jenkins', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175'})
    
    TASK [Create molecule instance(s)] *********************************************
    changed: [localhost] => (item={'name': 'jenkins', 'region': 'eu-west-2', 'image_owner': 334305766942, 'image_name': 'codeyourinfra/jenkins', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175'})
    
    TASK [Wait for instance(s) creation to complete] *******************************
    FAILED - RETRYING: Wait for instance(s) creation to complete (300 retries left).
    FAILED - RETRYING: Wait for instance(s) creation to complete (299 retries left).
    FAILED - RETRYING: Wait for instance(s) creation to complete (298 retries left).
    FAILED - RETRYING: Wait for instance(s) creation to complete (297 retries left).
    FAILED - RETRYING: Wait for instance(s) creation to complete (296 retries left).
    FAILED - RETRYING: Wait for instance(s) creation to complete (295 retries left).
    changed: [localhost] => (item={'started': 1, 'finished': 0, 'ansible_job_id': '111813117773.309', 'results_file': '/home/gcarmo/.ansible_async/111813117773.309', '_ansible_parsed': True, 'changed': True, '_ansible_no_log': False, 'failed': False, 'item': {'name': 'jenkins', 'region': 'eu-west-2', 'image_owner': 334305766942, 'image_name': 'codeyourinfra/jenkins', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175'}, 'index': 0, '_ansible_item_result': True, '_ansible_ignore_errors': None, '_ansible_item_label': {'name': 'jenkins', 'region': 'eu-west-2', 'image_owner': 334305766942, 'image_name': 'codeyourinfra/jenkins', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175'}})
```

ssbarnea commented:
How can I trigger tests only on AWS? I want to help testing this change manually but I do not know how to run molecule tests with aws only.

gustavomcarmo commented:
Hi @decentral1se! I believe the [Travis CI job](https://travis-ci.com/ansible/molecule/jobs/188525256) failed for some reason nothing to do with my changes...

decentral1se commented:
Restarted @gustavomcarmo! You're right :)

gustavomcarmo commented:
@decentral1se I'm thinking about getting always the latest image, by making a change like [this](https://github.com/codeyourinfra/codeyourinfra/commit/3c9fa57b98345b6b9afd425b706072ef00d007e2).

This change will enable using wildcards in `image_name`. For example, if `ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*` is set, the latest image of Ubuntu bionic is got.

What do you think about it?

gustavomcarmo commented:
@decentral1se [Here](https://github.com/codeyourinfra/codeyourinfra/blob/master/unarchive_from_url_param/molecule/aws/molecule.yml) is an example of `molecule.yml` configured for getting the **latest Ubuntu bionic EC2 image**:

```yml
platforms:
  - name: repo
    region: eu-west-2
    image_owner: 334305766942
    image_name: codeyourinfra/repo
    instance_type: t2.micro
    vpc_subnet_id: subnet-1cb17175
  - name: server1
    region: eu-west-2
    image_owner: 099720109477
    image_name: ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*
    instance_type: t2.micro
    vpc_subnet_id: subnet-1cb17175
    groups:
      - servers
  - name: server2
    region: eu-west-2
    image_owner: 099720109477
    image_name: ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*
    instance_type: t2.micro
    vpc_subnet_id: subnet-1cb17175
    groups:
      - servers
```

decentral1se commented:
> How can I trigger tests only on AWS? I want to help testing this change manually but I do not know how to run molecule tests with aws only.

Well, I just looked and actually, we have **no** functional AWS tests at all. I better make a ticket ...

> What do you think about it?

Sounds fine to me!

decentral1se commented:
>> How can I trigger tests only on AWS? I want to help testing this change manually but I do not know how to run molecule tests with aws only.
>> 
> Well, I just looked and actually, we have no functional AWS tests at all. I better make a ticket ...

Or actually, I won't be chasing this AWS test implementation so I leave it up to someone else.

gustavomcarmo commented:
@decentral1se Just pushed the **get the latest image** feature.

Some logs from my project:

```
    TASK [Get the image ami] *******************************************************
    ok: [localhost] => (item={'name': 'repo', 'region': 'eu-west-2', 'image_owner': 334305766942, 'image_name': 'codeyourinfra/repo', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175'})
    ok: [localhost] => (item={'name': 'server1', 'region': 'eu-west-2', 'image_owner': '099720109477', 'image_name': 'ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175', 'groups': ['servers']})
    ok: [localhost] => (item={'name': 'server2', 'region': 'eu-west-2', 'image_owner': '099720109477', 'image_name': 'ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175', 'groups': ['servers']})
    
    TASK [Create molecule instance(s)] *********************************************
    changed: [localhost] => (item={'name': 'repo', 'region': 'eu-west-2', 'image_owner': 334305766942, 'image_name': 'codeyourinfra/repo', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175'})
    changed: [localhost] => (item={'name': 'server1', 'region': 'eu-west-2', 'image_owner': '099720109477', 'image_name': 'ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175', 'groups': ['servers']})
    changed: [localhost] => (item={'name': 'server2', 'region': 'eu-west-2', 'image_owner': '099720109477', 'image_name': 'ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175', 'groups': ['servers']})
    
    TASK [Wait for instance(s) creation to complete] *******************************
    FAILED - RETRYING: Wait for instance(s) creation to complete (300 retries left).
    FAILED - RETRYING: Wait for instance(s) creation to complete (299 retries left).
    FAILED - RETRYING: Wait for instance(s) creation to complete (298 retries left).
    changed: [localhost] => (item={'started': 1, 'finished': 0, 'ansible_job_id': '993155834401.16208', 'results_file': '/home/gcarmo/.ansible_async/993155834401.16208', '_ansible_parsed': True, 'changed': True, '_ansible_no_log': False, 'failed': False, 'item': {'name': 'repo', 'region': 'eu-west-2', 'image_owner': 334305766942, 'image_name': 'codeyourinfra/repo', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175'}, 'index': 0, '_ansible_item_result': True, '_ansible_ignore_errors': None, '_ansible_item_label': {'name': 'repo', 'region': 'eu-west-2', 'image_owner': 334305766942, 'image_name': 'codeyourinfra/repo', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175'}})
    FAILED - RETRYING: Wait for instance(s) creation to complete (300 retries left).
    FAILED - RETRYING: Wait for instance(s) creation to complete (299 retries left).
    changed: [localhost] => (item={'started': 1, 'finished': 0, 'ansible_job_id': '373723558160.16236', 'results_file': '/home/gcarmo/.ansible_async/373723558160.16236', '_ansible_parsed': True, 'changed': True, '_ansible_no_log': False, 'failed': False, 'item': {'name': 'server1', 'region': 'eu-west-2', 'image_owner': '099720109477', 'image_name': 'ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175', 'groups': ['servers']}, 'index': 1, '_ansible_item_result': True, '_ansible_ignore_errors': None, '_ansible_item_label': {'name': 'server1', 'region': 'eu-west-2', 'image_owner': '099720109477', 'image_name': 'ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175', 'groups': ['servers']}})
    changed: [localhost] => (item={'started': 1, 'finished': 0, 'ansible_job_id': '249531352563.16262', 'results_file': '/home/gcarmo/.ansible_async/249531352563.16262', '_ansible_parsed': True, 'changed': True, '_ansible_no_log': False, 'failed': False, 'item': {'name': 'server2', 'region': 'eu-west-2', 'image_owner': '099720109477', 'image_name': 'ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175', 'groups': ['servers']}, 'index': 2, '_ansible_item_result': True, '_ansible_ignore_errors': None, '_ansible_item_label': {'name': 'server2', 'region': 'eu-west-2', 'image_owner': '099720109477', 'image_name': 'ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*', 'instance_type': 't2.micro', 'vpc_subnet_id': 'subnet-1cb17175', 'groups': ['servers']}})
```

How can I help with the AWS test implementation?

decentral1se commented:
> How can I help with the AWS test implementation?

Thanks for offering but be careful ... there be dragons ;) 

It would be something like:

* 1) (unit tests): Adding a `test_aws.py` in https://github.com/ansible/molecule/tree/master/test/unit/driver
* 2) (functional tests part 1): Adding a `aws` folder with plays in https://github.com/ansible/molecule/tree/master/test/resources/playbooks
* 3) (functional tests part 2): Adding a `aws` molecule config in https://github.com/ansible/molecule/tree/master/test/scenarios/driver
* 4) (functional tests part 3): Adding `aws` tests to https://github.com/ansible/molecule/blob/master/test/functional/test_command.py

Should you (or anyone) undertake such a mission you will surely be revered in the Molecule hall of fame.

:man_astronaut: :woman_astronaut:
